File: task_3.pl
read_lines_from_file(File, Lines) :-
  read_file_to_string(File, Content, []),
  split_string(Content, "\n", "", Lines).

parse_transition(Line, transition(FromState, ToState, Symbol)) :-
  split_string(Line, " ", "", [FromState, ToState, Symbol]).

generate_all_strings_helper(FiniteAutomaton, NextState, CurrentString, K) :-
  string_length(CurrentString, K),
  member(NextState, FiniteAutomaton.finalStates),
  write(CurrentString), write(' ').

generate_all_strings_helper(FiniteAutomaton, CurrentState, CurrentString, K) :-
  string_length(CurrentString, Length),
  Length < K,
  member(transition(CurrentState, NextState, Symbol), FiniteAutomaton.transitions),
  string_concat(CurrentString, Symbol, NextString),
  generate_all_strings_helper(FiniteAutomaton, NextState, NextString, K).

generate_all_strings(FiniteAutomaton, K) :-
  generate_all_strings_helper(FiniteAutomaton, FiniteAutomaton.startState, "", K),
  fail.

generate_all_strings(_, _).

main :-
  read_lines_from_file('states.txt', States),
  read_lines_from_file('symbols.txt', Symbols),
  read_lines_from_file('final_states.txt', FinalStates),
  read_lines_from_file('transitions.txt', TransitionsContent),
  read_lines_from_file('k.txt', [KContent]),

  maplist(parse_transition, TransitionsContent, Transitions),

  nth0(0, States, StartState),

  Automaton = _{states:States, symbols:Symbols, transitions:Transitions, startState:StartState, finalStates:FinalStates},

  number_string(K, KContent),

  generate_all_strings(Automaton, K).

:- initialization(main).

File: task_3.ts
interface Transition {
  from: string;
  to: string;
  symbol: string;
}

class FiniteAutomaton {
  states: string[];
  symbols: string[];
  transitions: Transition[];
  startState: string;
  finalStates: string[];

  constructor(
    states: string[],
    symbols: string[],
    transitions: Transition[],
    startState: string,
    finalStates: string[]
  ) {
    this.states = states;
    this.symbols = symbols;
    this.transitions = transitions;
    this.startState = startState;
    this.finalStates = finalStates;
  }

  generateAllStringsHelper(
    currentState: string,
    currentString: string,
    allStrings: string[],
    k: number
  ) {
    if (currentString.length <= k) {
      if (
        currentString.length === k &&
        this.finalStates.includes(currentState)
      ) {
        allStrings.push(currentString);
      }

      if (currentString.length < k) {
        for (const transition of this.transitions) {
          if (transition.from === currentState) {
            const nextState = transition.to;
            const symbol = transition.symbol;
            this.generateAllStringsHelper(
              nextState,
              currentString + symbol,
              allStrings,
              k
            );
          }
        }
      }
    }
  }

  generateAllStrings(k: number) {
    const allStrings: string[] = [];
    this.generateAllStringsHelper(this.startState, "", allStrings, k);
    return allStrings;
  }
}

const states = ["q0", "q1", "q2", "q3"];

const symbols = ["a", "b", "c"];

const startState = states[0];

const transitions = [
  { from: "q0", to: "q1", symbol: symbols[0] },
  { from: "q0", to: "q2", symbol: symbols[1] },
  { from: "q0", to: "q3", symbol: symbols[2] },

  { from: "q1", to: "q0", symbol: symbols[0] },
  { from: "q1", to: "q2", symbol: symbols[1] },
  { from: "q1", to: "q3", symbol: symbols[2] },

  { from: "q2", to: "q1", symbol: symbols[0] },
  { from: "q2", to: "q0", symbol: symbols[1] },
  { from: "q2", to: "q3", symbol: symbols[2] },

  { from: "q3", to: "q3", symbol: symbols[2] },
  { from: "q3", to: "q2", symbol: symbols[1] },
  { from: "q3", to: "q1", symbol: symbols[0] },
];

const finalStates = [states[3]];

const automaton = new FiniteAutomaton(
  states,
  symbols,
  transitions,
  startState,
  finalStates
);

const k = 3;
console.log(automaton.generateAllStrings(k));
